File: FacebookImageCachingService.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.IO;

public static class FacebookImageCachingService
{
  private class CachedImage
  {
    public byte[] Bytes;
    public string Url;
  }

  private static CachedImage _cachedImage;

  public static async Task<byte[]> GetImageAsync(string facebookPictureUrl)
  {
    if (_cachedImage == null)
    {
      if (File.Exists(Constants.FacebookImageCacheFileName))
      {
        Console.WriteLine($"Loading cached image from {Constants.FacebookImageCacheFileName}");
        try
        {
          var text = File.ReadAllText(Constants.FacebookImageCacheFileName);
          _cachedImage = JsonConvert.DeserializeObject<CachedImage>(text, new JsonSerializerSettings { DateParseHandling = DateParseHandling.None });
        }
        catch (Exception ex)
        {
          Console.WriteLine($"Error reading {Constants.FacebookImageCacheFileName}; will ignore file. Error was: " + ex);
          _cachedImage = new CachedImage();
        }
      }
      else
      {
        Console.WriteLine($"No cached image exists at {Constants.FacebookImageCacheFileName}");
        _cachedImage = new CachedImage();
      }
    }

    if (_cachedImage.Url == facebookPictureUrl)
    {
      Console.WriteLine($"Found facebook image in local cache");
      return _cachedImage.Bytes;
    }

    _cachedImage.Bytes = await DownloadFacebookImageAsync(facebookPictureUrl);
    _cachedImage.Url = facebookPictureUrl;

    Console.WriteLine($"Saving downloaded facebook image to cache file");
    File.WriteAllText(Constants.FacebookImageCacheFileName, JsonConvert.SerializeObject(_cachedImage, Formatting.Indented));

    return _cachedImage.Bytes;
  }

  private static async Task<byte[]> DownloadFacebookImageAsync(string url)
  {
    Console.WriteLine("Downloading image from facebook: " + url);
    using (var pictureStream = new MemoryStream())
    using (var client = new HttpClient())
    {
      HttpResponseMessage response = await client.GetAsync(url);
      using var responseStream = await response.Content.ReadAsStreamAsync();
      responseStream.CopyTo(pictureStream);
      pictureStream.Position = 0;
      if (response.IsSuccessStatusCode)
      {
        Console.WriteLine($"facebook's response: success ({pictureStream.Length} bytes)");
        return pictureStream.ToArray();
      }
      else
      {
        using var resultReader = new StreamReader(pictureStream);
        var result = resultReader.ReadToEnd();
        Console.WriteLine("facebook's response: " + result);
        throw new Exception($"Picture download request failed with response {(int)response.StatusCode} ({response.StatusCode}) {response.ReasonPhrase}");
      }
    }
  }
}

File: FacebookImageCachingService.js
const Constants = require('./Constants.js').Constants;
const fsPromises = require("fs").promises;

class FacebookImageCachingService {}

let _cachedImage = null;

// returns Promise<Uint8Array> of file contents, or async throws exception if it can't get it
FacebookImageCachingService.GetImageAsync = async (facebookPictureUrl) =>
{
  if (_cachedImage === null)
  {
    const statResult = await fsPromises.stat(Constants.FacebookImageCacheFileName);
    if (statResult)
    {
      console.log(`Loading cached image from ${Constants.FacebookImageCacheFileName}`);
      try
      {
        var jsonText = await fsPromises.readFile(Constants.FacebookImageCacheFileName, 'utf8');
        _cachedImage = JSON.parse(jsonText);
        if (!_cachedImage || !_cachedImage.bytes || !_cachedImage.bytes.length || !_cachedImage.url)
        {
          console.log(`Invalid content in ${Constants.FacebookImageCacheFileName}; will ignore file`);
          _cachedImage = {};
        }
      }
      catch (ex)
      {
        console.log(`Error reading ${Constants.FacebookImageCacheFileName}; will ignore file. Error was: ${ex}`);
        _cachedImage = {};
      }
    }
    else
    {
      console.log(`No cached image exists at ${Constants.FacebookImageCacheFileName}`);
      _cachedImage = {};
    }
  }

  if (_cachedImage.url === facebookPictureUrl)
  {
    console.log(`Found facebook image in local cache`);
    return _cachedImage.bytes;
  }

  _cachedImage.bytes = await FacebookImageCachingService.DownloadFacebookImageAsync(facebookPictureUrl);
  _cachedImage.url = facebookPictureUrl;

  console.log(`Saving downloaded facebook image to cache file`);
  const cachedImageText = JSON.stringify(_cachedImage, null, 2);
  await fsPromises.writeFile(Constants.FacebookImageCacheFileName, cachedImageText, 'utf8');

  return _cachedImage.bytes;
};

// returns Promise<Uint8Array> of file contents, or async throws exception if it can't get it
FacebookImageCachingService.DownloadFacebookImageAsync = async (url) =>
{
  console.log("Downloading image from facebook: " + url);

  const response = await fetch(url);
  if (!response.ok) {
    throw `fetch failed with status=${response.status} (${response.statusText})`;
  }
  const data = new Uint8Array(await response.arrayBuffer());
  return data;
};

module.exports = {
   FacebookImageCachingService: FacebookImageCachingService
}